### PR TITLE
Feature/bearing selection

### PR DESCRIPTION
**Pull request by danpat**
This adds a new `b=` parameter to `match` and `viaroute` requests which restricts Nearest Neighbor selection when finding phantom nodes.

If bearing hints are available to the user (i.e. from vehicle GPS traces), this can help get better snapping by not selecting road lanes going in the opposite direction to what was actually being travelled.

The parameter is optional, and if omitted, the current behaviour remains unchanged.

Like the `timestamp` parameter, if present, one `b` parameter is required for each coordinate pair.

The filter limits are currently hard-coded, but could probably be made configurable.  The intent at this stage is mostly to reduce matching edges going opposite to that which is desired, so a fairly broad matching angle is permissible.

The filter is implemented in the StaticRTree, so this could be trivially exposed for `/nearest`, and all other APIs that provide coordinates.

### CODE REVIEW

File: data_structures/route_parameters.cpp [hunk 1]
```diff
@@ -117,6 +117,15 @@ void RouteParameters::addTimestamp(const unsigned timestamp)
...
 }
 
+void RouteParameters::addBearing(const int bearing)
+{
+    bearings.resize(coordinates.size());
+    if (!bearings.empty())
+    {
+        bearings.back() = bearing;
+    }
+}
```
Comment (daniel-j-h) [inline data_structures/route_parameters.cpp]: Let me see if I understand this correctly: this is the handler for the `b=` spirit grammar, called with ints. It implicitly relies on interleaving coordinates and bearings. In the case of multiple bearings such as `b=1&b=2&b=3` only the last `b=3` is stored.

File: data_structures/static_rtree.hpp [hunk 2]
```diff
@@ -642,11 +642,56 @@ class StaticRTree
...
+    /**
+     * Checks whether A is between B-range and B+range, all modulo 360
+     * e.g. A = 5, B = 5, range = 10 == true
+     *      A = -6, B = 5, range = 10 == false
+     *      A = -2, B = 355, range = 10 == true
+     *      A = 6, B = 355, range = 10 == false
+     *      A = 355, B = -2, range = 10 == true
+     *
+     * @param A the bearing to check, in degrees, 0-359, 0=north
+     * @param B the bearing to check against, in degrees, 0-359, 0=north
```
Comment (daniel-j-h) [inline data_structures/static_rtree.hpp]: The documentation states as precondition for `A` and `B` to be in the range 0-359 (both including I suppose?).
But then you do modulo calculation to bring `A` and `B` in this range in case it is outside of it.

Following that:

- either the precondition is wrong and I am allowed to pass in values outside of the stated range, or
- the precondition is right, then assert this instead of doing modulo calculation

File: data_structures/static_rtree.hpp [hunk 3]
```diff
@@ -642,11 +642,56 @@ class StaticRTree
...
+     * @param range the number of degrees either side of B that A will still match
+     * @return true if B-range <= A <= B+range, modulo 360
+     * */
+    static bool IsBearingWithinBounds(const int A,
+                                      const int B,
+                                      const int range)
```
Comment (daniel-j-h) [inline data_structures/static_rtree.hpp]: Right now `range` is known at compile-time. Will this ever change?

If not tell the compiler this useful information, such as:

    template <int Range>
    fn() {
        static_assert(range >= 0,"");
        //
    }

Also: if the range is never allowed to be negative, then either assert this everywhere or use an integral unsigned type.

File: data_structures/static_rtree.hpp [hunk 4]
```diff
@@ -642,11 +642,56 @@ class StaticRTree
...
+
+
+
     bool IncrementalFindPhantomNodeForCoordinate(
         const FixedPointCoordinate &input_coordinate,
         std::vector<PhantomNode> &result_phantom_node_vector,
         const unsigned max_number_of_phantom_nodes,
         const float max_distance = 1100,
+        const int filter_bearing = 0,
+        const int filter_bearing_range = 180,
```
Comment (daniel-j-h) [inline data_structures/static_rtree.hpp]: In case you change the range argument to be a compile-time constant, you have to say `const constexpr int` here.

File: plugins/match.hpp [hunk 5]
```diff
@@ -126,8 +127,12 @@ template <class DataFacadeT> class MapMatchingPlugin : public BasePlugin
             }
 
             std::vector<std::pair<PhantomNode, double>> candidates;
+            // Use bearing values if supplied, otherwise fallback to 0,180 defaults
+            int bearing = input_bearings.size() > 0 ? input_bearings[current_coordinate] : 0;
+            int range = input_bearings.size() > 0 ? 10 : 180;
```
Comment (daniel-j-h) [inline plugins/match.hpp]: Those can be `const`. Also most variables _could_ be `auto` ..

File: plugins/match.hpp [hunk 6]
```diff
@@ -261,12 +266,19 @@ template <class DataFacadeT> class MapMatchingPlugin : public BasePlugin
...
             json_result.values["status"] = "Number of timestamps does not match number of coordinates .";
             return 400;
         }
 
+        if (input_bearings.size() > 0 && input_coords.size() != input_bearings.size())
+        {
+            json_result.values["status"] = "Number of bearings does not match number of coordinates .";
+            return 400;
+        }
+
```
Comment (daniel-j-h) [inline plugins/match.hpp]: I see how this checks at least that the sizes are the same. But what happens if I do not interleave coordinated with bearings, but instead do a request such as:

> coordinate, bearing, coordinate, coordinate, bearing, bearing

then this check won't catch the mistake and from the last two bearings, only the last one will be used and there exists a bearing that is default initialized but never assigned to?

File: data_structures/route_parameters.cpp (hunk 1 above)
Comment (danpat) [inline data_structures/route_parameters.cpp]: Correct, I've just mimicked the behaviour of `timestamp`.  In theory we could supply bearings only on some coordinates, but I didn't dig that far into the logic for now.

File: data_structures/static_rtree.hpp (hunk 2 above)
Comment (danpat) [inline data_structures/static_rtree.hpp]: Yep.  I'm being restrictive in the documentation, but forgiving in the implementation.  Because of edge cases around 0 degrees, it turned out to be easier to implement negative values anyway.  Docs are just to encourage users to keep it to 0-359.

File: server/data_structures/internal_datafacade.hpp [hunk 7]
```diff
@@ -447,13 +448,15 @@ template <class EdgeDataT> class InternalDataFacade final : public BaseDataFacad
         }
 
         return m_static_rtree->IncrementalFindPhantomNodeForCoordinate(
-            input_coordinate, resulting_phantom_node_vector, number_of_results);
+            input_coordinate, resulting_phantom_node_vector, number_of_results, 1100, bearing, range);
```
Comment (daniel-j-h) [inline server/data_structures/internal_datafacade.hpp]: Meh, it's bad enough that we hard-code the max. distance, but having to hard-code it multiple times is just ugly.

File: server/data_structures/shared_datafacade.hpp [hunk 8]
```diff
@@ -418,13 +419,15 @@ template <class EdgeDataT> class SharedDataFacade final : public BaseDataFacade<
         }
 
         return m_static_rtree->second->IncrementalFindPhantomNodeForCoordinate(
-            input_coordinate, resulting_phantom_node_vector, number_of_results);
+            input_coordinate, resulting_phantom_node_vector, number_of_results, 1100, bearing, range);
```
Comment (daniel-j-h) [inline server/data_structures/shared_datafacade.hpp]: Same here with hard-coding the max. distance. I understand how this is required here for being able to pass bearing and range with values other than the default ones. But maybe reorder the arguments then?

File: server/data_structures/internal_datafacade.hpp (hunk 7 above)
Comment (danpat) [inline server/data_structures/internal_datafacade.hpp]: Arg, I cringed as I typed that, I meant to re-order those parameters so it didn't need to be specified.  I'll fix this.

File: unit_tests/data_structures/static_rtree.cpp [hunk 9]
```diff
@@ -487,4 +493,100 @@ BOOST_AUTO_TEST_CASE(rectangle_test)
...
+    BOOST_CHECK_EQUAL(false, TestStaticRTree::IsBearingWithinBounds(344, -5, 10));
+    BOOST_CHECK_EQUAL(false, TestStaticRTree::IsBearingWithinBounds(6, -5, 10));
+
+    BOOST_CHECK_EQUAL(true, TestStaticRTree::IsBearingWithinBounds(-1, 5, 10));
+    BOOST_CHECK_EQUAL(false, TestStaticRTree::IsBearingWithinBounds(-6, 5, 10));
+
+    BOOST_CHECK_EQUAL(true, TestStaticRTree::IsBearingWithinBounds(-721, 5, 10));
+    BOOST_CHECK_EQUAL(true, TestStaticRTree::IsBearingWithinBounds(719, 5, 10));
+
+}
```
Comment (daniel-j-h) [inline unit_tests/data_structures/static_rtree.cpp]: Hmm reading those tests I have to think about edge-cases like `std::numeric_limits<int>::max()`. It should be safe right now, but would make for great tests when the implementation changes, as we handle values from the user (outside) here.

**Comment by emiltin** [comment]
very nice.  for viaroute, can you pass a bearing just for A, not B?

File: data_structures/static_rtree.hpp [hunk 10]
```diff
@@ -737,10 +782,33 @@ class StaticRTree
...
+                    !IsBearingWithinBounds(backward_edge_bearing, filter_bearing, filter_bearing_range))
+                {
+                    continue;
+                }
+
                 // store phantom node in result vector
                 result_phantom_node_vector.emplace_back(current_segment,
                                                         foot_point_coordinate_on_segment);
 
+                if (!IsBearingWithinBounds(forward_edge_bearing, filter_bearing, filter_bearing_range))
```
Comment (TheMarex) [inline data_structures/static_rtree.hpp]: Can we cache the result from the first if above for this?

File: data_structures/static_rtree.hpp [hunk 11]
```diff
@@ -893,6 +978,15 @@ class StaticRTree
                         current_segment.forward_travel_mode, current_segment.backward_travel_mode),
                     current_perpendicular_distance);
 
+                if (!IsBearingWithinBounds(forward_edge_bearing, filter_bearing, filter_bearing_range))
```
Comment (TheMarex) [inline data_structures/static_rtree.hpp]: Same as in the other call, I think we should cache the result here.

File: plugins/match.hpp (hunk 6 above)
Comment (TheMarex) [inline plugins/match.hpp]: Actually this will work. What probably won't work would be `bearing, bearing, bearing, coordinate, coordinate, coordinate`. We should actually fix this on a grammar level.

File: server/data_structures/shared_datafacade.hpp (hunk 8 above)
Comment (TheMarex) [inline server/data_structures/shared_datafacade.hpp]: :+1: for re-ordering.

File: unit_tests/data_structures/static_rtree.cpp (hunk 9 above)
Comment (TheMarex) [inline unit_tests/data_structures/static_rtree.cpp]: We should handle validation of route parameters on the server level (e.g. `route_parameters.cpp`).

This is definitely already garbage-in garbage-out territory.

Comment (daniel-j-h) [inline unit_tests/data_structures/static_rtree.cpp]: Right, but not garbage-in, undefined-behavior-out.

File: data_structures/static_rtree.hpp (hunk 4 above)
Comment (danpat) [inline data_structures/static_rtree.hpp]: I foresee changing this at runtime at some future time, but not in this PR. I'll `constexpr` these.

File: unit_tests/data_structures/static_rtree.cpp (hunk 9 above)
Comment (danpat) [inline unit_tests/data_structures/static_rtree.cpp]: I added a range check to `route_parameters.cpp`.

File: data_structures/static_rtree.hpp [hunk 12]
```diff
@@ -642,10 +642,55 @@ class StaticRTree
...
+                                      const Angle B,
+                                      const int range)
+    {
+
+        if (range >= 180) return true;
+        if (range <= 0) return false;
+
+        // Map both bearings into positive modulo 360 space
+        const int normalized_B = (int(B) < 0)?(int(B) % 360)+360:(int(B) % 360);
+        const int normalized_A = (int(A) < 0)?(int(A) % 360)+360:(int(A) % 360);
```
Comment (daniel-j-h) [inline data_structures/static_rtree.hpp]: Please no C-style casts, a la `(int)`. Those will try all casts, with even falling back to `reinterpret_cast` (which is bad). Do a `static_cast` here, maybe cache its value.

Or better: strongly typedef-ing does not prevent you from extending the type.
You could still add functions, such as:

    inline int normalize(const Angle& a) { return /* my normalization code */; }

and then reuse that throughout the codebase. You can even put your typedef and those functions into their own namespace. With ADL you only have to say `normalize(myns::Angle(90))`, as normalize is being searched in the `msns` namespace automatically (this is ADL).

File: strong_typedef.hpp [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    struct D                                                        \
+        : boost::totally_ordered1< D                                \
+        , boost::totally_ordered2< D, T                             \
+        > >                                                         \
+    {                                                               \
+        T val;                                                      \
+        explicit D(const T t_) : val(t_) {};                        \
+        D(): val() {};                                              \
+        D(const D & t_) : val(t_.val){}                             \
+        D & operator=(const D & rhs) { val = rhs.val; return *this;}  \
```
Comment (daniel-j-h) [inline strong_typedef.hpp]: This prevents the compiler from generating the special member move functions. Argh, I'm not sure this Boost typedef thing is up to date at all (re. C++11), after looking at the details.

Reference:

- http://accu.org/content/conf2014/Howard_Hinnant_Accu_2014.pdf p28
